enhancements/cluster-constraints-properties.md

```diff
@@ -0,0 +1,172 @@
+
+## Motivation
+
+### Existing dependency resolution
+
+OLM currently only resolves constraints that are provided by the operators and the resolution outcome at runtime is solely based on the constraints being satisfied by other operators who provide the required properties. It is not possible at the moment for OLM to consider cluster information or properties during resolution or apply cluster constraints to the available operators to prevent or allow certain operators to be installed successfully.
+
+### Cluster information
+
+Operators can depend on certain information such as usable API(s) that are available in the cluster. It is important to surface those information so operators can determine at runtime if they will be able to install and work properly in the cluster. The cluster information/properties can be depended upon as a constraint for a particular operator. This dependency relationship will ensure only applicable operators can be installed in the cluster and prevent possible failed installation and potentially nonworking condition.
+
+### Cluster restriction
+
+The cluster itself can impose certain restriction to which operators can be installed in the cluster via the properties associated with the operators. For example, the cluster imposes a minumum kubernetes version all for installable operators. A constraint can be constructed for a specific cluster restriction and then will be resolved during dependency resolution to ensure only operators with corresponding property with a satisfiable value.
+
+### Goals
+
+- Provide the mechanism to provide cluster properties and constraints
+- Expose cluster properties and constraints to users
+
+### Non-Goals
+- Determine which cluster properties/constraints are required on the cluster
+- Provide mechanism to retrieve and/or set up cluster properties and constraints automatically
+- Provide mechanism to have separate namespaced and cluster-scoped properties/constraints
+
+## Proposal
+
+A new constraint API provides an effective away for cluster administrators to create custom resources (CR) to contain cluster information such as cluster constraints and properties.
+
+OLM resolver will retrieve the cluster constraints and properties from the CR(s) and the information can be used during resolution.
+
+The cluster constraints and properties are presented as a virtual cluster-scoped operator that is installed in the namespace. For all incoming operators that are intended to run the namespace, they must satisfy all constraints that are associated with this cluster-scoped operator. At the same time, the installable operators can depend on the properties that are provided by this virtual operator.
+
+### User Stories
+
+#### Cluster constraints
+
+As a cluster admin, I would like to specify constraints that are specified to the cluster and they that will be applied to all operators. As a result, only operators that satisfy these constraints to be installable in the cluster.
```

Comment:
As per my comment above. As a cluster-admin I would like primarily that operators get only installed on clusters where they can work on without me needing to do extra things for it. Do you have a concrete example of a scenario where a cluster-admin would want to define what I would call "reverse constraints"? Today they can leverage catalogs to make operators available or not on specific clusters.